core/src/main/scala/shapeless/interface.scala

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (c) 2016 Miles Sabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package shapeless
+
+import scala.language.experimental.macros
+
+import scala.reflect.macros.whitebox
+
+trait Interface[T] extends Serializable {
+  type Repr
+  def to(t : T) : Repr
+  def from(r: Repr): T
+}
+
+object Interface {
+  type Aux[T, Repr0] = Interface[T] { type Repr = Repr0 }
+  def apply[T](implicit iface: Interface[T]): Aux[T, iface.Repr] = iface
+  implicit def materialize[T, R]: Interface.Aux[T, R] =
+    macro InterfaceMacros.materializeUnlabelledInterface[T, R]
+}
+
+trait FlattenHList[Nested <: HList] {
```

Comment:
This is for end-user code, which the tests are mock examples of.

This typeclass is what I'm referring to in the initial PR description with:
> Currently, the ending of e.g. a method with shape (Int)(Int) => Int (two param lists) is (Int :: HNil) :: (Int :: HNil) :: HNil => Int, with there being typeclass instances for conversion to a flattened-param-list version (Int :: Int :: HNil => Int). The question is whether we care about keeping the param lists separate or whether the initial encoding should just be flat in the first place